### PR TITLE
[jecoute]dev download file stat question

### PR DESCRIPTION
**Pull request by Raulnet**

### CODE REVIEW

File: tests/jecoute/FileStatsServiceTest.php [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+
+    public function testGetFilenameCsv(): void
+    {
+        $survey = new Survey();
+        $survey->setName('test u');
+        $response = $this->invokeMethod($this->fileStatsService, 'getFilenameCsv', [$survey]);
+        $this->assertSame('test_u_'.date('Y-m-d_H-i').'.csv', $response);
+    }
+
+    private function invokeMethod(
```
Comment (Remg) [inline tests/jecoute/FileStatsServiceTest.php]: IMHO, we are not supposed to test private methods.
We only test the API of a class, aka the public methods.

No matter how the private methods behaves: only the API (public methods) will have an impact on the outside (the application)

So.. i think you can remove this hack and tests about privates methods, what do you think ?

File: src/Jecoute/FileStatsService.php [hunk 2]
```diff
@@ -0,0 +1,93 @@
+<?php
+
+namespace AppBundle\Jecoute;
+
+use AppBundle\Entity\Jecoute\Survey;
+use AppBundle\Entity\Jecoute\SurveyQuestion;
+use AppBundle\Repository\Jecoute\DataAnswerRepository;
+
+class FileStatsService
```
Comment (thomasbisignani) [inline src/Jecoute/FileStatsService.php]: What do you think to rename this class by `StatisticsExporter` ?

File: src/Jecoute/FileStatsService.php [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+    private $dataAnswerRepository;
+    private $content;
+
+    public function __construct(StatisticsProvider $provider, DataAnswerRepository $dataAnswerRepository)
+    {
+        $this->dataProvider = $provider;
+        $this->dataAnswerRepository = $dataAnswerRepository;
+    }
+
+    public function getStatsFileCsvBySurvey(Survey $survey): array
```
Comment (thomasbisignani) [inline src/Jecoute/FileStatsService.php]: What do you think to rename this method by `export()` ?

File: tests/jecoute/FileStatsServiceTest.php [hunk 4]
```diff
@@ -0,0 +1,154 @@
+<?php
+
+namespace Tests\AppBundle\jecoute;
```
Comment (Remg) [inline tests/jecoute/FileStatsServiceTest.php]: **J**ecoute*

same for the directory

File: tests/jecoute/FileStatsServiceTest.php (hunk 1 above)
Comment (Raulnet) [reply tests/jecoute/FileStatsServiceTest.php]: ok removed

File: tests/jecoute/FileStatsServiceTest.php (hunk 4 above)
Comment (Raulnet) [reply tests/jecoute/FileStatsServiceTest.php]: fixed

File: src/Jecoute/FileStatsService.php (hunk 2 above)
Comment (Raulnet) [reply src/Jecoute/FileStatsService.php]: renamed

File: src/Jecoute/FileStatsService.php (hunk 3 above)
Comment (Raulnet) [reply src/Jecoute/FileStatsService.php]: renamed

File: src/Controller/EnMarche/ReferentController.php [hunk 5]
```diff
@@ -353,4 +354,33 @@ public function editReferentPersonLink(Request $request, ReferentPersonLinkRepos
...
+     * @Security("is_granted('IS_AUTHOR_OF', survey)")
+     */
+    public function jecouteSurveyStatsDownloadAction(
+        Survey $survey,
+        StatisticsExporter $statisticsExporter
+    ): Response {
+        $dataFile = $statisticsExporter->export($survey);
+
+        $response = new Response($dataFile['content'], 200, [
+            'Content-Type' => 'text/plain',
```
Comment (ottaviano) [inline src/Controller/EnMarche/ReferentController.php]: `text/csv`

File: src/Controller/EnMarche/ReferentController.php [hunk 6]
```diff
@@ -353,4 +354,33 @@ public function editReferentPersonLink(Request $request, ReferentPersonLinkRepos
...
+            'Content-Disposition' => 'attachment;filename="'.$dataFile['filename'].'"',
+        ]);
+        $response->send();
```
Comment (ottaviano) [inline src/Controller/EnMarche/ReferentController.php]: why do we need this?

File: src/Controller/EnMarche/ReferentController.php [hunk 7]
```diff
@@ -353,4 +354,33 @@ public function editReferentPersonLink(Request $request, ReferentPersonLinkRepos
...
+     *
+     * @Security("is_granted('IS_AUTHOR_OF', survey)")
+     */
+    public function jecouteSurveyStatsDownloadAction(
+        Survey $survey,
+        StatisticsExporter $statisticsExporter
+    ): Response {
+        $dataFile = $statisticsExporter->export($survey);
+
+        $response = new Response($dataFile['content'], 200, [
```
Comment (ottaviano) [inline src/Controller/EnMarche/ReferentController.php]: `200` => const `Response::OK`

File: src/Jecoute/StatisticsExporter.php [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+    private function getContent(Survey $survey): string
+    {
+        $data = $this->dataProvider->getStatsBySurvey($survey);
+        $this->buildHeader($survey);
+
+        foreach ($data['questions'] as $question) {
+            $this->content .= $question['content'].\PHP_EOL;
+
+            switch ($question['type']) {
+                case SurveyQuestionTypeEnum::MULTIPLE_CHOICE_TYPE:
```
Comment (ottaviano) [inline src/Jecoute/StatisticsExporter.php]:
```php
case SurveyQuestionTypeEnum::MULTIPLE_CHOICE_TYPE:
case SurveyQuestionTypeEnum::UNIQUE_CHOICE_TYPE:
    $this->buildRowsChoicesAnswers($question['stats']);
    break;
case ...:
```

File: src/Jecoute/StatisticsExporter.php [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+
+            $this->content .= \PHP_EOL;
+        }
+
+        return $this->content;
+    }
+
+    private function buildHeader(Survey $survey): void
+    {
+        $this->content = $survey->getName().' '.date('Y-m-d H:i').\PHP_EOL.\PHP_EOL;
```
Comment (ottaviano) [inline src/Jecoute/StatisticsExporter.php]: `d/m/Y H:i`?

File: src/Jecoute/StatisticsExporter.php [hunk 10]
```diff
@@ -0,0 +1,93 @@
...
+    private function buildRowsSimpleField(SurveyQuestion $surveyQuestion): void
+    {
+        $responses = $this->dataAnswerRepository->findAllBySurveyQuestion($surveyQuestion);
+
+        if (!empty($responses)) {
+            foreach ($responses as $response) {
+                $postAt = $response['postedAt'];
+
+                if ($postAt instanceof \DateTimeInterface) {
+                    $response['postedAt'] = $postAt->format('Y-m-d H:i:s');
```
Comment (ottaviano) [inline src/Jecoute/StatisticsExporter.php]: `d/m/Y H:i`?

File: src/Controller/EnMarche/ReferentController.php (hunk 7 above)
Comment (Raulnet) [reply src/Controller/EnMarche/ReferentController.php]: remove

File: src/Controller/EnMarche/ReferentController.php (hunk 5 above)
Comment (Raulnet) [reply src/Controller/EnMarche/ReferentController.php]: edited

File: src/Controller/EnMarche/ReferentController.php (hunk 6 above)
Comment (Raulnet) [reply src/Controller/EnMarche/ReferentController.php]: removed

File: src/Jecoute/StatisticsExporter.php (hunk 8 above)
Comment (Raulnet) [reply src/Jecoute/StatisticsExporter.php]: done

File: src/Jecoute/StatisticsExporter.php (hunk 9 above)
Comment (Raulnet) [reply src/Jecoute/StatisticsExporter.php]: edited

File: src/Jecoute/StatisticsExporter.php (hunk 10 above)
Comment (Raulnet) [reply src/Jecoute/StatisticsExporter.php]: edited

**Pull request merged**